### DISCUSSION TITLE
Changing the query key used as params without using state

### DISCUSSION THREAD

**Discussion by dm0114** [TanStack/query#6704] [Q&A] [answered]
I've heard that the query key comparison only happens on render, is there any way to use the query key without using state, and if so, where should I refer to the relevant source code?

**Comment by TkDodo** [comment]
what do you mean "without state" ? The values in the QueryKey are the dependencies to your query, they have to come from somewhere. Usually it's user input or some other dynamic values that are essentially "client state". You have to manage that _somehow_

**Reply by dm0114** [reply]
Sorry, I meant state using useState, not refs or variable values.

**Reply by TkDodo** [reply]
show an example of what you mean please

**Reply by dm0114** [reply]
```jsx
export const useParams = () => {
  const selectRef = useRef<HTMLInputElement>(null);
  const [input, _setInput] = useDebouncedState<string | null>(null, 200);

  const params = {
    page: 1,
    size: 10,
    sort: 'id,desc',
    selection: selectRef.current?.value ?? undefined,
    input: input,
  };

  const handleChangeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
    _setInput(e.currentTarget.value);
  };

  const fetchNextPage = () => {
    params.page = params.page + 1
  }

  return {
    params,
    selectRef,
    handleChangeInput,
    fetchNextPage,
  }
}

export const useTestList = (params) => {
  return useQuery({
    queryKey: ['test', params],
    queryFn: queryFn()
  })
}

export function Page() {
  const {params, selectRef, handleChangeInput, fetchNextPage}  = useParams()
  const {data} = useTestList(params)
  return (
    <div>
      <select ref={selectRef} data={data}/> // not working
      <input defaultValue={params.input ?? ''} onChange={handleChangeInput} /> // working
      <button onClick={fetchNextPage}>+1</button> // not working
    </div>
  )
} 
```
I didn't need the state in that example and the documentation said  _As long as the query key is serializable, and unique to the query's data, you can use it!_, So I tried to change the queryKey using the object's reference, but it didn't work. I was looking through the documentation and Q&A and came across [this post](https://stackoverflow.com/questions/73295707/how-does-reactquery-monitor-complex-query-key-changes-by-value-and-or-reference) and stopped looking, but I was wondering if there was another way to do it.

**Reply by TkDodo** [reply, marked as answer]
```
const fetchNextPage = () => {
  params.page = params.page + 1
}
```

the rules of react still apply. Try to display `params` in a `<div>` and you would get the same "not working" state. You can't write to refs and expect them to magically update somewhere. That's what state is for ...

**Reply by dm0114** [reply]
Thank you for answer
